### PR TITLE
Improvements on Tesserae implementation

### PR DESCRIPTION
**Pull request by karljohanw**
-  Memory improvements: Only allocate space actually _used_ in matrices
- Performance improvements: Avoiding repeated calculations improves the speed with up to 22%

### CODE REVIEW

File: src/cortexpy/tesserae.py [hunk 1]
```diff
@@ -369,12 +365,12 @@ def __initialization(self, query, panel, lsize_l):
                     self.vt_i[seq][pos_seq][1] = self.lpiI - lsize_l + self.lsi[convert[query[0]]]
 
                     if pos_seq > 0:
-                        self.vt_d[seq][pos_seq][1] = self.vt_m[seq][pos_seq - 1][1] + self.ldel
-                        self.tb_d[seq][pos_seq][1] = seq * 10 + 1 + (pos_seq - 1) / self.tb_divisor
-
-                        if self.vt_d[seq][pos_seq - 1][1] + self.leps > self.vt_d[seq][pos_seq][1]:
-                            self.vt_d[seq][pos_seq][1] = self.vt_d[seq][pos_seq - 1][1] + self.leps
-                            self.tb_d[seq][pos_seq][1] = seq * 10 + 3 + (pos_seq - 1) / self.tb_divisor
+                        vt_d_1, tb_d_1 = max([
```
Comment (winni2k) [inline src/cortexpy/tesserae.py:368]: I don't think the brackets (`[` and `]`) are necessary here?

**Comment by winni2k** [comment]
Lots of nice code cleanup here. @kvg do you have any thoughts?

**Comment by karljohanw** [comment]
A few words concerning the memory improvements:

The matrices allocated (6 totally), when run with targets with maximum length of 37738bp, were of sizes 2×37738² (=number of target sequences×length of maximum target^2), although all of it isn't used. This would require ~137GB RAM. (As a comparation, a node on UPPMAX/rackham provides 128GB RAM.) By instead allocating the matrices: number of target sequences×length of maximum target×length of current query (this way it is allocated in the (supplement of) tesserae paper), one would be able to run a query of 20000bp on the same targets using ~72GB RAM.

**Comment by winni2k** [comment]
Thanks for the clarification @karljohanw!

**Comment by kvg** [comment]
Sorry for my delay - I’ve set aside some time for my review today.

Thanks,
-Kiran

File: src/cortexpy/tesserae.py (hunk 1 above)
Comment (kvg) [reply src/cortexpy/tesserae.py:368]: Agreed - max(...) has a function prototype that takes multiple arguments, removing the need to prepack everything into a tuple first.

File: src/cortexpy/tesserae.py [hunk 2]
```diff
@@ -75,14 +76,15 @@ def __init__(self, pdel=DEFAULT_DEL, peps=DEFAULT_EPS, prho=DEFAULT_REC, pterm=D
     def __initialize(self, query, targets):
         self.nseq = len(targets) + 1
         self.maxl = max_length(query, targets)
+        self.qlen = len(query)
 
-        self.vt_m = np.full([self.nseq, self.maxl + 1, self.maxl + 1], SMALL, dtype=np.float64)
-        self.vt_i = np.full([self.nseq, self.maxl + 1, self.maxl + 1], SMALL, dtype=np.float64)
-        self.vt_d = np.full([self.nseq, self.maxl + 1, self.maxl + 1], SMALL, dtype=np.float64)
+        self.vt_m = np.full([self.nseq, self.maxl + 1, self.qlen + 2], SMALL, dtype=np.float64)
```
Comment (kvg) [inline src/cortexpy/tesserae.py:81]: Simple yet effective change.  Perhaps the +1 to maxl and +2 to qlen should be moved into lines 78 and 79 to make them symmetric with the nseq + 1 in line 77.  Though I need to read on to see if that would screw anything up too badly.

Also, why is it qlen + 2 and not qlen + 1?

File: src/cortexpy/tesserae.py [hunk 3]
```diff
@@ -237,20 +246,11 @@ def __termination(self, l1, pos_max, state_max, who_max):
         pos_next = 0
         while pos_target >= 1:
             if state_max == 1:
-                who_next = int(self.tb_m[who_max][pos_max][pos_target] / 10)
-                state_next = int(self.tb_m[who_max][pos_max][pos_target] - who_next * 10)
-                pos_next = int((self.tb_m[who_max][pos_max][pos_target] - who_next * 10 - state_next) *
-                               self.tb_divisor + 1e-6)
+                who_next, state_next, pos_next = self.__to_traceback_indices(self.tb_m[who_max][pos_max][pos_target])
```
Comment (kvg) [inline src/cortexpy/tesserae.py:249]: Nice - much much cleaner!

**Review by kvg — APPROVED** [review]
This is really nice!  I'm delighted to see that this refactoring as such a huge effect on the memory consumption and overall performance of Tesserae.  Thanks for catching my error in the (ridiculously wasteful) matrix allocations.

Are there still some useful memory reductions to be obtained by storing only certain columns of the traceback and computing forward when necessary?  Ultimately my goal will be to treat ultra-long Nanopore reads (~0.1 Mb, maximum ~3 Mbp) with Tesserae.  But I'm not sure what the engineering requirements are to get there yet.

**Comment by kvg** [comment]
Unrelated to the improvements to Tesserae, there seems to be an issue with tests/test_unit/test_graph/test_unitig_search.py that's preventing the build tests from completing successfully.  It's happening on my laptop too.  Any ideas as to what's causing it?

**Comment by winni2k** [comment]
Hmm, I'll take a deeper look into why the tests aren't running properly. For now, I have checked out this branch and can confirm that all tests still pass on my laptop. Obviously, I want the tests to run on other people's machines and for the test environment to be easy to set up, so I'll have a look at this next week.

**Comment by winni2k** [comment]
I'll open a new issue for the HMM backtrack changes.

**Pull request merged**